### PR TITLE
Fix possible pipeline connections leak

### PR DESCRIPTION
**Pull request by ING-XIAOJIAN**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Do tests and lints pass with this change?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change
When Executing "n.write()" may generate some unknown errors(e.g. DataError), which could result in the connection not being released, finally lead to connection leak.

### CODE REVIEW

**Comment by dvora-h** [comment]
@ING-XIAOJIAN thanks for suggestion, looks good, but WDYT about:
```
    try:
        node_commands = nodes.values()
        for n in node_commands:
            n.write()

        for n in node_commands:
            n.read()
    finally:
        for n in nodes.values():
            n.connection_pool.release(n.connection)
```

**Comment by ING-XIAOJIAN** [comment]
> @ING-XIAOJIAN thanks for suggestion, looks good, but WDYT about:
> 
> ```
>     try:
>         node_commands = nodes.values()
>         for n in node_commands:
>             n.write()
> 
>         for n in node_commands:
>             n.read()
>     finally:
>         for n in nodes.values():
>             n.connection_pool.release(n.connection)
> ```
I naively think "n.read" never generates some unknown errors,
I will take your suggestion.

**Comment by dvora-h** [comment]
I agree that there is less chance to get an error in the reading, but I prefer to cover that as well just to be safe

**Comment by ING-XIAOJIAN** [comment]
> @ING-XIAOJIAN Can you fix the linters or wants me to do it?

I already fixed those linters

**Comment by dvora-h** [comment]
Thanks!

File: redis/cluster.py [hunk 1]
```diff
@@ -2190,8 +2195,8 @@ def _send_cluster_commands(
         # buffered result back from a previous command and
         # every single request after to that connection will always get
         # a mismatched result.
-        for n in nodes.values():
-            n.connection_pool.release(n.connection)
+        # for n in nodes.values():
```
Comment (dvora-h) [inline redis/cluster.py:2198]: Can you delete this lines from the comment and move the explanations of this lines before them?

Comment (ING-XIAOJIAN) [reply redis/cluster.py]: I already fixed this. could you review it again?

File: redis/cluster.py [hunk 2]
```diff
@@ -2166,32 +2166,34 @@ def _send_cluster_commands(
...
+            node_commands = nodes.values()
+            for n in node_commands:
+                n.write()
+
+            for n in node_commands:
+                n.read()
+        finally:
+            # release all of the redis connections we allocated earlier
+            # back into the connection pool.
+            # we used to do this step as part of a try/finally block,
```
Comment (bcmills) [inline redis/cluster.py:2179]: This comment seems out-of-date — isn't it explaining why _not_ to do exactly what this change is doing?

It should probably be update to explain why this _is_ safe, not why it _isn't_. 😅

**Pull request merged**